File: Features/Gatherings/PlayerCountRange.cs
namespace oyinQ.Bot.Features.Gatherings;

public readonly record struct PlayerCountRange(int Minimum, int Maximum, bool WasDefaulted)
{
    public const int DefaultMinimum = 1;
    public const int DefaultMaximum = 12;

    public static PlayerCountRange Normalize(int? minimum, int? maximum) =>
        minimum is >= 1 && maximum is >= 1 && minimum <= maximum
            ? new PlayerCountRange(minimum.Value, maximum.Value, false)
            : new PlayerCountRange(DefaultMinimum, DefaultMaximum, true);

    public PlayerCountRange WithExpansions(IEnumerable<oyinQ.Bot.Features.Collections.ClubCollectionExpansion> expansions)
    {
        var minimum = Minimum;
        var maximum = Maximum;
        foreach (var expansion in expansions)
        {
            var range = Normalize(expansion.MinPlayers, expansion.MaxPlayers);
            if (range.WasDefaulted) continue;
            minimum = Math.Min(minimum, range.Minimum);
            maximum = Math.Max(maximum, range.Maximum);
        }
        return new(minimum, maximum, WasDefaulted);
    }
}

File: MiniApp/src/pages/gatherings/playerCountRange.ts
export type PlayerCountRange = {
  minimum: number;
  maximum: number;
  wasDefaulted: boolean;
};

export function normalizePlayerCountRange(minimum?: number, maximum?: number): PlayerCountRange {
  if (Number.isInteger(minimum) && Number.isInteger(maximum)
      && minimum! >= 1 && maximum! >= minimum!) {
    return { minimum: minimum!, maximum: maximum!, wasDefaulted: false };
  }

  return { minimum: 1, maximum: 12, wasDefaulted: true };
}

export function resolvePlayerCountRange(minimum: number | undefined, maximum: number | undefined,
  expansions: Expansion[], selected: number[]): PlayerCountRange {
  const range = normalizePlayerCountRange(minimum, maximum);
  for (const expansion of expansions) {
    if (!selected.includes(expansion.bggId)) continue;
    const extra = normalizePlayerCountRange(expansion.minPlayers, expansion.maxPlayers);
    if (extra.wasDefaulted) continue;
    range.minimum = Math.min(range.minimum, extra.minimum);
    range.maximum = Math.max(range.maximum, extra.maximum);
  }
  return range;
}

export type PlayerLimits = { minimum: number; desired: number; maximum: number };

export function fitPlayerLimits(limits: PlayerLimits, before: PlayerCountRange, after: PlayerCountRange): PlayerLimits {
  const clamp = (value: number) => Math.max(after.minimum, Math.min(after.maximum, value));
  const minimum = clamp(limits.minimum);
  const maximum = limits.maximum === before.maximum ? after.maximum : clamp(limits.maximum);
  return { minimum, maximum, desired: Math.max(minimum, Math.min(maximum, limits.desired)) };
}
import type { Expansion } from "../../api/types";
